README
# Chef::Knife::Proxmox

Authors: Jorge Moratilla, Sergio Galvan
Date: 2012-10-07


## Description

This knife plugin allows to access Proxmox Virtualization Environment (Proxmox VE or PVE) through its API.
It allows you to lists templates installed on the server/cluster (like images or amis), create a server or
destroy it.

Currently it only supports management of openVZ instances, but KVM instances will be able in a future.

## Requirements

* zlib (zlib1g-dev)
* openssl (libopenssl-ruby, libssl-dev)

Then install your ruby language (recommended ruby-1.9.3-xxx)

(See installation instructions beneath)


## Actions implemented

+ proxmox server list
+ proxmox server info
+ proxmox server create
+ proxmox server start
+ proxmox server stop
+ proxmox server destroy
+ proxmox template available
+ proxmox template list

## Some Examples

### List servers
    $ knife proxmox server list
    Id   Node    Name                 Type    Status
    104  node-1  tetatet.example.com  openvz  down
    470  node-2  sg-node1.example.com openvz  up
    102  node-1  test                 qemu    down
    202  node-2  chef-client1         openvz  down

### Get info for a server
    $ knife proxmox server info -H tetatet.example.com -P ip
    192.168.1.1

    $ knife proxmox server info -H tetatet.example.com
    knife proxmox server info -H tetatet.example.com -P ip
    cpu:       0.00281876923420022
    cpus:      2
    disk:      5833015296
    diskread:  316895232
    diskwrite: 3805184
    failcnt:   0
    ha:        0
    ip:        192.168.42.200
    maxdisk:   8589934592
    maxmem:    4294967296
    maxswap:   536870912
    mem:       1145778176
    name:      tetatet.example.com
    netin:     27888604
    netout:    2829601
    nproc:     140
    status:    running
    swap:      54460416
    type:      openvz
    uptime:    61018


### List templates installed
    $ knife proxmox  template list -U https://localhost:8006/api2/json/ -n localhost -u test -p test123 -R pve -VV
    DEBUG: Using configuration from /home/jorge/workspace/chef-repo/.chef/knife.rb
    Id  Name                                                         Size
    0   local:vztmpl/debian-6.0-request-tracker_3.8.8-2_i386.tar.gz  171 MB
    1   local:vztmpl/old_ubuntu-11.10-x86_64.tar.gz                  124 MB
    2   local:vztmpl/ubuntu-10.04-standard_10.04-4_i386.tar.gz       135 MB
    3   local:vztmpl/ubuntu-11.10-x86_64-jorge1-.tar.gz              124 MB
    4   local:vztmpl/ubuntu-11.10-x86_64-jorge2-.tar.gz              154 MB


### List templates available to download
    $ knife proxmox template available -U https://localhost:8006/api2/json/ -u test -p test123 -n localhost -R pve  -VV
    DEBUG: Using configuration from /home/jorge/workspace/chef-repo/.chef/knife.rb
    Name                                                       Operating System
    debian-6-turnkey-concrete5_12.0-1_i386.tar.gz              debian-6
    ubuntu-10.04-turnkey-prestashop_11.3-1_i386.tar.gz         ubuntu-10.04
    debian-6-turnkey-joomla25_12.0-1_i386.tar.gz               debian-6
    debian-6-turnkey-tomcat-apache_12.0-1_i386.tar.gz          debian-6
    debian-6.0-wordpress_3.4.2-1_i386.tar.gz                   debian-6.0 .....


### Create a server (read the note at the end of the document.  It's about obtaining the IPAddress)

    $  knife proxmox server create -n ankh -r "recipe[java]" -C 2 -M 1024 -H example-server -P test123 -T 4
    Creating VM 473...
    ..............OK
    Preparing the server to start
    Starting VM 473 on node ankh....
    ..OK
    New Server 473 has IP Address: 10.0.2.19
    done
    Bootstrapping Chef on 10.0.2.19
    10.0.2.19 --2013-01-23 01:27:20--  http://opscode.com/chef/install.sh
    10.0.2.19 Resolving opscode.com...
    10.0.2.19 184.106.28.83
    10.0.2.19 Connecting to opscode.com|184.106.28.83|:80...
    10.0.2.19 connected.
    10.0.2.19 HTTP request sent, awaiting response...
    10.0.2.19 301 Moved Permanently
    10.0.2.19 Location: http://www.opscode.com/chef/install.sh [following]
    10.0.2.19 --2013-01-23 01:27:21--  http://www.opscode.com/chef/install.sh
    10.0.2.19 Resolving www.opscode.com...
    10.0.2.19 184.106.28.83
    10.0.2.19 Reusing existing connection to opscode.com:80.
    10.0.2.19 HTTP request sent, awaiting response...
    10.0.2.19 200 OK


### Starting a server
    $ knife proxmox server start --vmid 401
    Starting VM 401....
    Result: 200


### Stopping a server
    $ knife proxmox server stop --vmid 103
    Stoping VM 103....
    Result: 200


### Destroy a server
    $ knife proxmox server destroy -U https://localhost:8006/api2/json/ -u test -p test123 -n localhost -R pve -N vm-node1 -VV
    DEBUG: Using configuration from /home/jorge/workspace/chef-repo/.chef/knife.rb
    node to destroy: vm-node1 [vmid: 200]
    Continue? (Y/N) y
    Stopping VM 303....
    Result: 200
    ..............................
    Result: 200

## Installation

This gem must be installed on your system using the command

    $ gem install knife-proxmox

If you want to read values from the environment variables modify your knife.rb:

    knife[:pve_cluster_url] ||= "#{ENV['PVE_CLUSTER_URL']}"
    knife[:pve_user_name]   ||= "#{ENV['PVE_USER_NAME']}"
    knife[:pve_node_name]   ||= "#{ENV['PVE_NODE_NAME']}"
    knife[:pve_user_realm]  ||= "#{ENV['PVE_USER_REALM']}"

However I don't recommend to have the user password as a environment variable.
knife[:pve_user_password] = "#{ENV['PVE_CLUSTER_URL']}"

### Environment Variables

    PVE_CLUSTER_URL = https://<proxmox_server>:8006/api2/json/
    PVE_USER_NAME   = <user>
    PVE_PASSWORD_NAME = <password>
    PVE_USER_REALM  = <pam|pve|others>
    PVE_NODE_NAME   = <nodename>

If you want to alter some variable by command line then you can use the following options:
-n --node <node>   If in a cluster, this argument will specify which node of the cluster will do the action
-U --pve_cluster_url <http://localhost...> Will change the cluster to access to
-u --username <user>   Same for user
-p --password <pass>   Same for password
-r --realm    <realm>  Same for the realm used to authenticate

### ABOUT CHEF AND PROXMOX

To bootstrap a newly created VM, you should apply the patch inside the extra directory.  Otherwise you must
specify the IP address in the knife proxmox server create --ipaddress argument.



CHANGELOG
Author: Jorge Moratilla
Date: 2012-10-07

[2013-05-07]
Version: 0.0.18
* knife proxmox server list now sorts output by vmid (default) and by field (vmid, node, name)

[2013-04-29]
Version: 0.0.17
* knife proxmox server destroy now requests confirmation to delete a server

[2013-01-23]
Version: 0.0.10
* knife proxmox server create can bootstrap a server, enabled by default

[2013-01-21]
Version: 0.0.9
* Added IP field to server list command

[2012-12-21]
Version: 0.0.8
* Added a patch for proxmox that allows me to fetch the IP of the guess
* Working with the PVE API

[2012-12-13]
Version: 0.0.6
* Added new field in proxmox server list to show the node the vm is in
* Now you can start, stop, destroy a server without to specify the node

[2012-12-12]
Version: 0.0.5
* Improved problems with selection of the node of the cluster
* Better checking of task status
* Solved problem #3 catching from exceptions

[2012-11-12]

Added an option to knife proxmox server create that allows the user to specify the ipaddress.  If not, then the VM will be created with DHCP enabled.
Created a very basic task control for creating and deleting servers.  Great refactoring needed.
Now, I'm using bitbucket issue control center to keep control of issues.  Please use this link to add control keys in pushes 
https://confluence.atlassian.com/display/BITBUCKET/Automatically+Resolving+Issues+when+Users+Push+Code 

[2012-11-06]

Added server start and stop commands
$ knife proxmox server start --vmid 401
Starting VM 401....
Result: 200

$ knife proxmox server stop --vmid 103
Stoping VM 103....
Result: 200


[2012-11-05]

First refactor, ProxmoxBase has now all the common parameters
Added proxmox server list to show current servers
Example:
$ knife proxmox server list
Id   Name                 Type    Status
104  tetatet.jnlabs.isp   openvz  down  
470  sg-node1.jnlabs.isp  openvz  up  
102  test                 qemu    down  
202  chef-client1         openvz  down 

[2012-10-24]

Improved knife commands.  Needs refactor internally
Examples:

List templates installed
$ knife proxmox  template list -U https://localhost:8006/api2/json/ -n localhost -u test -p test123 -R pve -VV
DEBUG: Using configuration from /home/jorge/workspace/chef-repo/.chef/knife.rb
Id  Name                                                         Size  
0   local:vztmpl/debian-6.0-request-tracker_3.8.8-2_i386.tar.gz  171 MB
1   local:vztmpl/old_ubuntu-11.10-x86_64.tar.gz                  124 MB
2   local:vztmpl/ubuntu-10.04-standard_10.04-4_i386.tar.gz       135 MB
3   local:vztmpl/ubuntu-11.10-x86_64-jorge1-.tar.gz              124 MB
4   local:vztmpl/ubuntu-11.10-x86_64-jorge2-.tar.gz              154 MB
 

List templates available to download
$ knife proxmox template available -U https://localhost:8006/api2/json/ -u test -p test123 -n localhost -R pve  -VV
DEBUG: Using configuration from /home/jorge/workspace/chef-repo/.chef/knife.rb
Name                                                       Operating System
debian-6-turnkey-concrete5_12.0-1_i386.tar.gz              debian-6        
ubuntu-10.04-turnkey-prestashop_11.3-1_i386.tar.gz         ubuntu-10.04    
debian-6-turnkey-joomla25_12.0-1_i386.tar.gz               debian-6        
debian-6-turnkey-tomcat-apache_12.0-1_i386.tar.gz          debian-6        
debian-6.0-wordpress_3.4.2-1_i386.tar.gz                   debian-6.0 .....


Create a server
$ knife proxmox server create -U https://localhost:8006/api2/json/ -n localhost -u test -p test123 -R pve -N vm-node1 --template 4 -VV 
DEBUG: Using configuration from /home/jorge/workspace/chef-repo/.chef/knife.rb
DEBUG: vmid=303&hostname=vm-node1&storage=local&password=pve123&ostemplate=local%3Avztmpl%2Fubuntu-11.10-x86_64-jorge2-.tar.gz&memory=512&swap=512&disk=4&cpus=1&netif=ifname%3Deth0%2Cbridge%3Dvmbr0
Result: 200
..............................
Starting VM 303....
Result: 200

Destroy a server
$ knife proxmox server destroy -U https://localhost:8006/api2/json/ -u test -p test123 -n localhost -R pve -N vm-node1 -VV
DEBUG: Using configuration from /home/jorge/workspace/chef-repo/.chef/knife.rb
node to destroy: vm-node1
Stopping VM 303....
Result: 200
..............................
Result: 200


[2012-10-17]

First preliminar version is working, but only as separate knife plugin files.

In order to work, these files must be copied to your ~/.chef/plugins/knife directory.  If it
doesn't exist, then you must create it.

Actions implemented

proxmox server create
proxmox server destroy
proxmox template available
proxmox template list

Next steps

R.E.F.A.C.T.O.R

Options in command line for each action


[2012-10-07]

This project is a plugin for knife that allows a user to list templates, create and delete 
proxmox virtual machines (openvz instances mainly)

I've two ways to do this:
A) Create a knife plugin directly using command line tools (pvesh and pvesm)
B) Create a provider for fog (github.com/fog/fog) and then create the knife plugin based on fog

Thoughts:
Option A is the easy quick and dirty solution for me now.
Option B is the proper way to do it.  Unfortunately, I'm not able to afford it.  It's too complex.

Actions:
Option A)

. knife proxmox template list
|- pvesh get /nodes/SERVER_NAME/storage/STORAGE/content
|--{
      "content" : "vztmpl",
      "format" : "tgz",
      "size" : 161655654,
      "volid" : "local:vztmpl/ubuntu-11.10-x86_64-jorge2-.tar.gz"
   },

. knife proxmox server create
|- pvesh create /nodes/localhost/openvz -ostemplate local:vztmpl/ubuntu-11.10-x86_64-jorge2-.tar.gz -vmid 106 -cpus 1 -memory 1024 -disk 4 -hostname jorge_test

. knife proxmox server destroy
|- pvesh destroy 106



samples/ProxmoxAPITest.rb
# Proxmox API Tests
# 
# Author: Jorge Juan Moratilla Porras
# 
# Description: some tests with the Proxmox VE API


#require 'rubygems'
require 'rest_client'
require 'json'
require 'pp'
require 'colorize'

# Default values
username = ENV['PVE_USER_NAME'] || 'test'
realm    = ENV['PVE_REALM'] || 'pve'
password = ENV['PVE_USER_PASSWORD'] || 'test123'
url_base = ENV['PVE_CLUSTER_URL'] || 'https://localhost:8006/'
nodename = ENV['PVE_NODE_NAME'] || 'localhost'
storagename = ENV['PVE_STORAGE_NAME'] || 'local'
url_base += 'api2/json/'

csrf_prevention_token = nil
token = nil

# RestClient logger
log = RestClient.log = []

puts "
SERVER INFO: 
  PVE_CLUSTER_URL  => #{url_base}
  PVE_NODE_NAME    => #{nodename}
  PVE_USER_NAME    => #{username}
  PVE_REALM        => #{realm}
  PVE_STORAGE_NAME => #{storagename}
"

site = RestClient::Resource.new(url_base)

puts 'AUTH'.blue

print ' To request access: '.yellow
site['access/ticket'].post :username=>username,:realm=>realm,:password=>password do |response, request, result, &block| 
  if response.code == 200 then
    data = JSON.parse(response.body)
    ticket = data['data']['ticket']
    csrf_prevention_token = data['data']['CSRFPreventionToken']
    if !ticket.nil? then
# Token is a cookie coded like this one
# 'PVEAuthCookie=PVE%3Atest@pve%3A5079E676%3A%3AE5Btg[...]crcp/RzEitO/vKMvr5YpAmjBRw7HS2IA3Q%3D%3D'
      token = 'PVEAuthCookie=' + ticket.gsub!(/:/,'%3A').gsub!(/=/,'%3D')
    end
  end
  puts "#{response.code}" 
end 

auth_params = {
  :CSRFPreventionToken => csrf_prevention_token,
  :cookie => token
}

puts 'GET'.blue

print ' list all users in the cluster: '.yellow
site['access/users'].get auth_params do |response, request, result, &block|
  puts "#{response.code}"
end

print ' list cluster resources: '.yellow
site['cluster/resources?type=vm'].get auth_params do |response, request, result, &block|
  puts "#{response.code}"
  data = JSON.parse(response.body)['data']
  data.each {|entry|
    puts "vmid: #{entry['vmid']}, host: #{entry['name']}, type: #{entry['type']}" #if entry['type'] == 'openvz' or entry['type'] == 'qemu'
  }
end

print ' list available templates for download: '.yellow
site["nodes/#{nodename}/aplinfo"].get auth_params do |response, request, result, &block|
  puts "#{response.code}"
  data = JSON.parse(response.body)['data']
  data.each { |entry|
    puts "#{entry['template']}: #{entry['os']}"
  }
end

print ' list all installed templates on a specific node: '.yellow
site["nodes/#{nodename}/storage/#{storagename}/content"].get auth_params do |response, request, result, &block|
  puts "#{response.code}"
  data = JSON.parse(response.body)['data']
  data.each {|entry|
    puts "#{entry['volid']}: #{entry['size']}" if entry['content'] == 'vztmpl'
  }
end

vm_id = 201
print ' list all server attributes for vmid = 201: '.yellow
site["nodes/#{nodename}/openvz/#{vm_id}/config"].get auth_params do |response, request, result, &block|
  puts "#{response.code}"
  data = JSON.parse(response.body)['data']
  puts "#{nodename}:"
  data.keys.each {|entry|
    puts "\t#{entry}: #{data[entry]}" if entry == 'netif'
  }
end

print ' list all VM\'s per node: '.yellow
site["nodes/#{nodename}/openvz"].get auth_params do |response, request, result, &block|
  puts "#{response.code}"
  data = JSON.parse(response.body)['data']
  puts "#{nodename}:"
  data.each {|entry|
    puts "\t#{entry['vmid']}: #{entry['name']}"
  }
end


print ' list all tasks in the cluster: '.yellow
site["cluster/tasks"].get auth_params do |response, request, result, &block|
  puts "#{response.code}" 
end

puts 'POST'.blue

print ' create a new user: '.yellow
site['access/users'].post 'userid=test2@pve', auth_params do |response, request, result, &block|
  puts "#{response.code}"
end

taskid = nil

print ' download a template: '.yellow
site["nodes/#{nodename}/aplinfo"].post "storage=local&template=ubuntu-10.04-turnkey-appengine_11.3-1_i386.tar.gz", auth_params do |response, request, result, &block|
  puts "#{response.code}"
  taskid = JSON.parse(response.body)['data'] # For last test (stop a running task)
end

print ' create an openvz VM: '.yellow
site["nodes/#{nodename}/openvz"].post 'vmid=401&hostname=melon1&storage=local&password=melon123&ostemplate=local%3Avztmpl%2Fubuntu-11.10-x86_64-jorge2-.tar.gz&memory=512&swap=512&disk=4&cpus=1&netif=ifname%3Deth0%2Cbridge%3Dvmbr0', auth_params do |response, request, result, &block|
  puts "#{response.code}"
end

print ' create an openvz VM without vmid: '.yellow
site["nodes/#{nodename}/openvz"].post 'hostname=melon2&storage=local&password=melon123&ostemplate=local%3Avztmpl%2Fubuntu-11.10-x86_64-jorge2-.tar.gz&memory=512&swap=512&disk=4&cpus=1&netif=ifname%3Deth0%2Cbridge%3Dvmbr0', auth_params do |response, request, result, &block|
  puts "#{response.code}"
end

puts 'PUT'.blue

print ' modify one user: '.yellow
site['access/users/test@pve'].put 'comment=hello world', auth_params do |response, request, result, &block|
  puts "#{response.code}"
end

puts "Sleeping 10 secs before deleting stuff"
sleep 10

puts 'DELETE'.blue

print ' destroy an existing user: '.yellow
site['access/users/test2@pve'].delete auth_params do |response, request, result, &block|
  puts "#{response.code}"
end

print ' destroy an existing openvz VM: '.yellow
site["nodes/#{nodename}/openvz/401"].delete auth_params do |response, request, result, &block|
  puts "#{response.code}"
end

print ' stop a running task: '.yellow
site["nodes/#{nodename}/tasks/#{taskid}"].delete auth_params do |response, request, result, &block|
  puts "#{response.code}"
end
